iquip/apps/builder.py

```diff
@@ -309,11 +312,83 @@ def __init__(
 
         Args:
             name: See the attributes section.
-            argInfo: See the attributes section.
+            argInfo: Each key and its value are:
+              default: The dictionary that describes arguments of a scannable object.
+              unit: The unit of the number value.
+              scale: The scale factor that is multiplied to the number value.
+              global_step: The step between values changed by the up and down button.
+              global_min: The minimum value. (default=float("-inf"))
+              global_max: The maximum value. (default=float("inf"))
+                If min > max, then they are swapped.
+              ndecimals: The number of displayed decimals.
         """
         super().__init__(parent=parent)
         self.name = name
-        self.argInfo = argInfo
+        self.state = self.get_state(argInfo)
+        self.stack = QStackedWidget(self)
+        self.layout = QGridLayout(self)
+        self.layout.addWidget(QLabel(name, self), 0, 0)
+        self.layout.addWidget(self.stack, 1, 1)
+
+    def get_state(self, argInfo):
+        """Gets a state dictionary that describes key parameters of a scannable object.
+
+        Creates a default state dictionary and updates it using the argInfo.
+
+        Args:
+            argInfo: A dictionary that contains argument information of each scannable type.
+        """
+        scale = argInfo["scale"]
+        state = {
+            "selected": "NoScan",
+            "NoScan": {"value": 0.0, "repetitions": 1},
+            "RangeScan": {"start": 0.0, "stop": 100.0*scale, "npoints": 10,
+                          "randomize": False, "seed": None},
+            "CenterScan": {"center": 0.*scale, "span": 100.*scale,
+                           "step": 10.*scale, "randomize": False,
+                           "seed": None},
+            "ExplicitScan": {"sequence": []}
+        }
+        if "default" in argInfo:
+            defaults = argInfo["default"]
+            if not isinstance(defaults, list):
+                defaults = [defaults]
+            state["selected"] = defaults[0]["ty"]
+            for default in reversed(defaults):
```

Comment:
Why did you reverse `defaults`?

Also, I think this for loop can be shorten as follows:
```python
for default in defaults:
  ty = default.pop("ty")
  if ty not in ["NoScan", "RangeScan", "CenterScan", "ExplicitScan"]:
    logger.warning("Unknown scan type: %s", ty)
  state[ty] = default
```

Reply:
I just used artiq code for reverse, but I also think it is meaningless.